FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,675 @@
+            IStorage storage,
+            IOptions<JobConfiguration> jobConfiguration,
+            ILogger<AzureStorageQueueClient<TJobInfo>> logger)
+        {
+            EnsureArg.IsNotNull(jobConfiguration, nameof(jobConfiguration));
+            EnsureArg.IsNotNull(storage, nameof(storage));
+
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+
+            if (storage.UseConnectionString)
+            {
+                _azureJobInfoTableClient = new TableClient(storage.TableUrl, storage.TableName);
+                _azureJobMessageQueueClient = new QueueClient(storage.QueueUrl, storage.QueueName);
+            }
+            else
+            {
+                _azureJobInfoTableClient = new TableClient(
+                    new Uri(storage.TableUrl),
+                    storage.TableName,
+                    new DefaultAzureCredential());
+                _azureJobMessageQueueClient = new QueueClient(
+                    new Uri($"{storage.QueueUrl}{storage.QueueName}"),
+                    new DefaultAzureCredential());
+            }
+        }
+
+        public bool IsInitialized()
+        {
+            try
+            {
+                _azureJobInfoTableClient.CreateIfNotExists();
+                _azureJobMessageQueueClient.CreateIfNotExists();
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to initialize azure storage client.");
+                return false;
+            }
+
+            _logger.LogInformation("Initialize azure storage client successfully.");
+            return true;
+        }
+
+        // The expected behaviors:
+        // 1. multi agent instances enqueue same jobs concurrently, only one job entity will be created,
+        //    there may be multi messages, while only one will be recorded in job lock entity, and the created jobInfo will be returned for all instances.
+        // 2. re-enqueue job, no matter what the job status is now, will do nothing, and return the existing jobInfo.
+        // 3. if one of the steps fails, will continue to process it when re-enqueue.
+        // TODO: The parameter forceOneActiveJobGroup and isCompleted are ignored for now
+        public async Task<IEnumerable<JobInfo>> EnqueueAsync(
+            byte queueType,
+            string[] definitions,
+            long? groupId,
+            bool forceOneActiveJobGroup,
+            bool isCompleted,
+            CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Start to enqueue {definitions.Length} jobs.");
+
+            // step 1: get incremental job ids, will try again if fails
+            var jobIds = await GetIncrementalJobIds(queueType, definitions.Length, cancellationToken);
+
+            // handle for each job
+            var jobInfos = new List<TJobInfo>();
+            for (var i = 0; i < definitions.Length; i++)
+            {
+                Trace.Assert(
```

Comment:
Do we need to throw a certain kind of exception here is Assert fail?

Reply:
If assert fails, the process will terminate.

By the way, without assert, if the entity is larger than 1MB, will throw  `413` exception with ErrorCode [RequestBodyTooLarge](https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes) when updating to the table.